pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,275 @@
+	// RegisterMetric registers a Prometheus Gauge or Counter metric for telemetry visitor.
+	// It must be called after creating the telemetry engine.
+	RegisterMetric(columnName string, gaugeOrCounterMetric prometheus.Metric) error
+	Close() error
+}
+
+func NewTelemetryEngine(conn pgxconn.PgxConn, uuid [16]byte) (Telemetry, error) {
+	// Warn the users about telemetry collection.
+	log.Warn("msg", "Promscale collects anonymous usage telemetry data for project improvements while being GDPR compliant. "+
+		"If you wish not to participate, please turn off sending telemetry in TimescaleDB. "+
+		"Details about the process can be found at https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry/#disabling-telemetry")
+	engine := &engine{
+		conn: conn,
+		uuid: uuid,
+	}
+	if err := engine.writeMetadata(); err != nil {
+		return nil, fmt.Errorf("writing metadata: %w", err)
+	}
+	return engine, nil
+}
+
+// writeMetadata writes Promscale and Tobs metadata. Must be written only by
+func (t *engine) writeMetadata() error {
+	promscale, err := promscaleMetadata()
+	if err != nil {
+		return fmt.Errorf("promscale metadata: %w", err)
+	}
+	promscale["promscale_exec_platform"] = ExecPlatform
+	if err := syncTimescaleMetadataTable(t.conn, Stats(promscale)); err != nil {
+		return fmt.Errorf("writing metadata for promscale: %w", err)
+	}
+
+	tobs := tobsMetadata()
+	if len(tobs) > 0 {
+		if err := syncTimescaleMetadataTable(t.conn, Stats(tobs)); err != nil {
+			return fmt.Errorf("writing metadata for tobs: %w", err)
+		}
+	}
+	return nil
+}
+
+// syncTimescaleMetadataTable syncs the metadata/stats with telemetry metadata table and returns the last error if any.
+func syncTimescaleMetadataTable(conn pgxconn.PgxConn, m Stats) error {
+	batch := conn.NewBatch()
+	for key, metadata := range m {
+		safe := pgutf8str.Text{}
+		if err := safe.Set(metadata); err != nil {
+			return fmt.Errorf("setting in pgutf8 safe string: %w", err)
+		}
+		query := "INSERT INTO _timescaledb_catalog.metadata VALUES ( $1, $2, true ) ON CONFLICT (key) DO UPDATE SET value = $2, include_in_telemetry = true WHERE metadata.key = $1"
+		batch.Queue(query, key, safe)
+	}
+	results, err := conn.SendBatch(context.Background(), batch)
+	if err != nil {
+		return fmt.Errorf("error sending batch: %w", err)
+	}
+	rows, err := results.Query()
+	if err != nil {
+		return fmt.Errorf("error querying results: %w", err)
+	}
+	for rows.Next() {
+		if err != nil {
+			err = rows.Err()
+		}
+		rows.Close()
+	}
+	return err
+}
+
+func (t *engine) Close() error {
+	if t.stopTelemetryRoutine != nil {
+		t.stopTelemetryRoutine()
+	}
+	if t.wgTelemetryRoutine != nil {
+		t.wgTelemetryRoutine.Wait()
+	}
+	return nil
+}
+
+func (t *engine) StartRoutineAsync() {
+	ctx, stopTelemetryRoutine := context.WithCancel(context.Background())
+	go t.telemetrySync(ctx)
+	t.stopTelemetryRoutine = stopTelemetryRoutine
+
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	t.wgTelemetryRoutine = wg
```

Comment:
I would assign to `t.wgTelemetryRoutine` before starting the goroutine since `telemetrySync` accesses it. I know it's not actually possible to call Done on the context before the `wgTelemetryRoutine` gets assigned, but it feels like a matter of good practice

Reply:
same for `stopTelemetryRoutine` actually